src/backend/farmer/maybe_node_client.rs

```diff
@@ -1,127 +1,708 @@
+        &self,
+        solution_response: SolutionResponse,
+    ) -> Result<(), Error> {
+        self.load_inner()?
+            .submit_solution_response(solution_response)
+            .await
+    }
+
+    async fn subscribe_slot_info(
+        &self,
+    ) -> Result<Pin<Box<dyn Stream<Item = SlotInfo> + Send + 'static>>, Error> {
+        self.load_inner()?.subscribe_slot_info().await
+    }
+
+    async fn submit_reward_signature(
+        &self,
+        reward_signature: RewardSignatureResponse,
+    ) -> Result<(), Error> {
+        self.load_inner()?
+            .submit_reward_signature(reward_signature)
+            .await
+    }
+
+    async fn subscribe_reward_signing(
+        &self,
+    ) -> Result<Pin<Box<dyn Stream<Item = RewardSigningInfo> + Send + 'static>>, Error> {
+        self.load_inner()?.subscribe_reward_signing().await
+    }
+
+    async fn subscribe_archived_segment_headers(
+        &self,
+    ) -> Result<Pin<Box<dyn Stream<Item = SegmentHeader> + Send + 'static>>, Error> {
+        self.load_inner()?
+            .subscribe_archived_segment_headers()
+            .await
     }
 
     async fn acknowledge_archived_segment_header(
         &self,
         segment_index: subspace_core_primitives::SegmentIndex,
     ) -> Result<(), Error> {
-        match &*self.inner.load() {
-            Some(inner) => {
-                inner
-                    .acknowledge_archived_segment_header(segment_index)
-                    .await
-            }
-            None => Err("Inner node client not injected yet".into()),
-        }
+        self.load_inner()?
+            .acknowledge_archived_segment_header(segment_index)
+            .await
+    }
+
+    async fn piece(
+        &self,
+        requested_piece_index: PieceIndex,
+    ) -> Result<Option<subspace_core_primitives::Piece>, Error> {
+        self.load_inner()?.piece(requested_piece_index).await
+    }
+
+    async fn segment_headers(
+        &self,
+        segment_indexes: Vec<SegmentIndex>,
+    ) -> Result<Vec<Option<subspace_core_primitives::SegmentHeader>>, Error> {
+        self.load_inner()?.segment_headers(segment_indexes).await
     }
 }
 
 #[async_trait::async_trait]
-impl NodeClientExt for MaybeNodeRpcClient {
+impl NodeClientExt for MaybeNodeClient {
     async fn last_segment_headers(&self, limit: u64) -> Result<Vec<Option<SegmentHeader>>, Error> {
-        match &*self.inner.load() {
-            Some(inner) => inner.last_segment_headers(limit).await,
-            None => Err("Inner node client not injected yet".into()),
-        }
+        self.load_inner()?.last_segment_headers(limit).await
     }
 }
 
-impl MaybeNodeRpcClient {
-    pub(in super::super) fn inject(&self, inner: NodeRpcClient) {
+impl MaybeNodeClient {
+    /// Load inner `NodeClient` with error handling
+    fn load_inner(&self) -> Result<Arc<Box<dyn NodeClientExt>>, Error> {
+        self.inner
+            .load()
+            .clone()
```

Comment:
Are you sure cloning is necessary here? We're not taking ownership fo the value, code worked without cloning just fine in the past.